File: app/src/main/java/com/example/myapplication/SignalProcessor.kt
package com.example.myapplication

import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.PI

/**
 * 信号处理器
 * FFT分析、幅度提取、滤波等
 */
class SignalProcessor {
    
    companion object {
        private const val SAMPLE_RATE = 48000  // 采样率 48 kHz（匹配脚本配置）
        private const val NUM_FREQUENCIES = 10  // 频率点数量
        private const val START_FREQ = 20000   // 起始频率 17.5kHz
        private const val FREQ_STEP = 350      // 频率步进 350Hz
        
        // 预计算二阶巴特沃斯高通滤波器系数（截止频率 15kHz）
        private val fc = 15000.0
        private val omega = 2.0 * PI * fc / SAMPLE_RATE
        private val cosOmega = cos(omega)
        private val sinOmega = sin(omega)
        
        // 二阶巴特沃斯滤波器 (Q = 0.7071)
        private val q = 0.7071
        private val alpha = sinOmega / (2.0 * q)
        private val b0 = (1.0 + cosOmega) / 2.0
        private val b1 = -(1.0 + cosOmega)
        private val b2 = (1.0 + cosOmega) / 2.0
        private val a0 = 1.0 + alpha
        private val a1 = -2.0 * cosOmega
        private val a2 = 1.0 - alpha
        
        // 归一化系数
        private val b0_norm = b0 / a0
        private val b1_norm = b1 / a0
        private val b2_norm = b2 / a0
        private val a1_norm = a1 / a0
        private val a2_norm = a2 / a0
    }
    
    // 高通滤波器状态变量
    private var hpf_x1 = 0.0
    private var hpf_x2 = 0.0
    private var hpf_y1 = 0.0
    private var hpf_y2 = 0.0
    
    /**
     * 二阶巴特沃斯高通滤波器 - 去除人声和低频噪声
     * 截止频率: 15kHz (人声通常 < 4kHz)
     * 优化：使用预计算的系数避免每次重复计算
     */
    fun highPassFilter(data: ShortArray): ShortArray {
        if (data.isEmpty()) return data
        
        val filtered = ShortArray(data.size)
        
        for (i in data.indices) {
            val x0 = data[i].toDouble()
            val y0 = b0_norm * x0 + b1_norm * hpf_x1 + b2_norm * hpf_x2 -
                     a1_norm * hpf_y1 - a2_norm * hpf_y2
            hpf_x2 = hpf_x1
            hpf_x1 = x0
            hpf_y2 = hpf_y1
            hpf_y1 = y0
            filtered[i] = y0.coerceIn(Short.MIN_VALUE.toDouble(), Short.MAX_VALUE.toDouble()).toInt().toShort()
        }
        
        return filtered
    }
    
    /**
     * 重置滤波器状态（开始新的录音时调用）
     */
    fun resetFilter() {
        hpf_x1 = 0.0
        hpf_x2 = 0.0
        hpf_y1 = 0.0
        hpf_y2 = 0.0
    }
    
    /**
     * 提取多个频率的幅度（使用 Goertzel 算法，比 FFT 更高效）
     */
    fun extractMagnitudes(audioData: ShortArray): FloatArray {
        val magnitudes = FloatArray(NUM_FREQUENCIES)
        
        for (i in 0 until NUM_FREQUENCIES) {
            val frequency = START_FREQ + i * FREQ_STEP
            magnitudes[i] = goertzelMagnitude(audioData, frequency)
        }
        
        // 【调试】每100帧打印一次幅度
        debugCounter++
        if (debugCounter % 100 == 0) {
            val max = magnitudes.maxOrNull() ?: 0f
            val avg = magnitudes.average().toFloat()
            android.util.Log.d("SignalProcessor", "幅度: 最大=${String.format("%.2f", max)}, 平均=${String.format("%.2f", avg)}")
        }
        
        return magnitudes
    }
    
    private var debugCounter = 0
    
    /**
     * Goertzel 算法 - 单频率检测
     * 比完整 FFT 更高效，适合只检测少数几个频率的场景
     */
    private fun goertzelMagnitude(samples: ShortArray, targetFreq: Int): Float {
        val n = samples.size
        val k = (0.5 + (n * targetFreq) / SAMPLE_RATE.toDouble()).toInt()
        val omega = (2.0 * PI * k) / n
        val coeff = 2.0 * cos(omega)
        
        var q0 = 0.0
        var q1 = 0.0
        var q2 = 0.0
        
        // Goertzel 迭代
        for (sample in samples) {
            q0 = coeff * q1 - q2 + sample.toDouble()
            q2 = q1
            q1 = q0
        }
        
        // 计算幅度
        val real = q1 - q2 * cos(omega)
        val imag = q2 * sin(omega)
        val magnitude = sqrt(real * real + imag * imag)
        
        return (magnitude / n).toFloat()  // 归一化
    }
    
    /**
     * 带通滤波器（简单的移动平均）
     */
    fun bandPassFilter(data: ShortArray): ShortArray {
        if (data.size < 3) return data
        
        val filtered = ShortArray(data.size)
        filtered[0] = data[0]
        
        for (i in 1 until data.size - 1) {
            // 简单的 3 点移动平均
            filtered[i] = ((data[i - 1].toInt() + data[i].toInt() + data[i + 1].toInt()) / 3).toShort()
        }
        
        filtered[data.size - 1] = data[data.size - 1]
        return filtered
    }
    
    /**
     * 计算 RMS (均方根) 能量
     */
    fun calculateRMS(data: ShortArray): Float {
        if (data.isEmpty()) return 0f
        
        var sum = 0.0
        for (sample in data) {
            sum += sample.toDouble() * sample.toDouble()
        }
        
        return sqrt(sum / data.size).toFloat()
    }
    
    /**
     * 计算平均幅度
     */
    fun calculateAverageMagnitude(magnitudes: FloatArray): Float {
        if (magnitudes.isEmpty()) return 0f
        return magnitudes.average().toFloat()
    }

    /**
     * Short数组转Float数组（归一化到[-1, 1]）
     */
    fun shortToFloatArray(data: ShortArray): FloatArray {
        return FloatArray(data.size) { i -> data[i] / 32768f }
    }

    /**
     * Teager-Kaiser包络提取
     * 使用公式: Ψ[x(n)] = x²(n) - x(n-1)·x(n+1)
     * 返回包络信号 = sqrt(|Ψ|)
     */
    fun teagerEnvelope(signal: FloatArray): FloatArray {
        val envelope = FloatArray(signal.size)
        // 边界点保持为0
        for (i in 1 until signal.size - 1) {
            val psi = signal[i] * signal[i] - signal[i - 1] * signal[i + 1]
            envelope[i] = sqrt(kotlin.math.abs(psi))
        }
        return envelope
    }

    /**
     * 计算Teager包络特征值（平均幅度）
     */
    fun calculateTeagerFeature(envelope: FloatArray): Float {
        if (envelope.isEmpty()) return 0f
        var sum = 0f
        var count = 0
        // 跳过边界点
        for (i in 1 until envelope.size - 1) {
            sum += envelope[i]
            count++
        }
        return if (count > 0) sum / count else 0f
    }
}

File: app/src/main/python/sd_core/signal_processor.py
"""
信号处理器
FFT分析、幅度提取、滤波等
"""
import numpy as np
from typing import Tuple
import math

class SignalProcessor:
	SAMPLE_RATE: int = 48000
	NUM_FREQUENCIES: int = 10
	START_FREQ: int = 20000
	FREQ_STEP: int = 350

	_fc: float = 15000.0
	_omega: float = 2.0 * math.pi * _fc / SAMPLE_RATE
	_cos_omega: float = math.cos(_omega)
	_sin_omega: float = math.sin(_omega)
	_q: float = 0.7071
	_alpha: float = _sin_omega / (2.0 * _q)
	_b0: float = (1.0 + _cos_omega) / 2.0
	_b1: float = -(1.0 + _cos_omega)
	_b2: float = (1.0 + _cos_omega) / 2.0
	_a0: float = 1.0 + _alpha
	_a1: float = -2.0 * _cos_omega
	_a2: float = 1.0 - _alpha

	b0_norm: float = _b0 / _a0
	b1_norm: float = _b1 / _a0
	b2_norm: float = _b2 / _a0
	a1_norm: float = _a1 / _a0
	a2_norm: float = _a2 / _a0

	def __init__(self):
		self._hpf_x1: float = 0.0
		self._hpf_x2: float = 0.0
		self._hpf_y1: float = 0.0
		self._hpf_y2: float = 0.0
		self._debug_counter: int = 0

	def high_pass_filter(self, data: np.ndarray) -> np.ndarray:
		"""
		二阶巴特沃斯高通滤波器 - 去除人声和低频噪声
		截止频率: 15kHz (人声通常 < 4kHz)

		Args:
			data: 输入音频数据 (ShortArray)

		Returns:
			滤波后的音频数据
		"""
		if len(data) == 0:
			return data

		filtered = np.zeros(len(data), dtype=np.int16)

		for i in range(len(data)):
			x0 = float(data[i])
			y0 = (self.b0_norm * x0 + self.b1_norm * self._hpf_x1 +
				self.b2_norm * self._hpf_x2 - self.a1_norm * self._hpf_y1 -
				self.a2_norm * self._hpf_y2)
			self._hpf_x2 = self._hpf_x1
			self._hpf_x1 = x0
			self._hpf_y2 = self._hpf_y1
			self._hpf_y1 = y0
			y0_clamped = max(-32768.0, min(32767.0, y0))
			filtered[i] = int(y0_clamped)

		return filtered

	def reset_filter(self) -> None:
		"""重置滤波器状态"""
		self._hpf_x1 = 0.0
		self._hpf_x2 = 0.0
		self._hpf_y1 = 0.0
		self._hpf_y2 = 0.0

	def extract_magnitudes(self, audio_data: np.ndarray) -> np.ndarray:
		"""
		提取多个频率的幅度（使用 Goertzel 算法，比 FFT 更高效）

		Args:
			audio_data: 音频数据

		Returns:
			各频率的幅度数组
		"""
		return self._vectorized_goertzel(audio_data)

	def _vectorized_goertzel(self, samples: np.ndarray) -> np.ndarray:
		"""
		使用 FFT 替代 Goertzel 算法，提高多频率检测效率

		Args:
			samples: 音频样本

		Returns:
			各频率的幅度数组
		"""
		n = len(samples)
		frequencies = np.array([self.START_FREQ + i * self.FREQ_STEP for i in range(self.NUM_FREQUENCIES)], dtype=np.float32)

		# 计算每个频率对应的 k 值
		k = (0.5 + (n * frequencies) / self.SAMPLE_RATE).astype(int)

		# 使用 FFT 计算所有频率的幅度
		fft_result = np.fft.fft(samples.astype(np.float64))
		magnitude = np.abs(fft_result[k]) / n

		return magnitude.astype(np.float32)

	def band_pass_filter(self, data: np.ndarray) -> np.ndarray:
		"""
		带通滤波器（简单的移动平均）

		Args:
			data: 输入音频数据

		Returns:
			滤波后的音频数据
		"""
		if len(data) < 3:
			return data

		filtered = np.zeros(len(data), dtype=np.int16)
		filtered[0] = data[0]

		for i in range(1, len(data) - 1):
			filtered[i] = int((int(data[i - 1]) + int(data[i]) + int(data[i + 1])) / 3)

		filtered[-1] = data[-1]
		return filtered

	@staticmethod
	def calculate_rms(data: np.ndarray) -> float:
		"""
		计算 RMS (均方根) 能量

		Args:
			data: 音频数据

		Returns:
			RMS值
		"""
		if len(data) == 0:
			return 0.0

		sum_val = 0.0
		for sample in data:
			sum_val += float(sample) ** 2

		return float(math.sqrt(sum_val / len(data)))

	@staticmethod
	def calculate_average_magnitude(magnitudes: np.ndarray) -> float:
		"""
		计算平均幅度

		Args:
			magnitudes: 幅度数组

		Returns:
			平均幅度
		"""
		if len(magnitudes) == 0:
			return 0.0
		return float(np.mean(magnitudes))

	@staticmethod
	def short_to_float_array(data: np.ndarray) -> np.ndarray:
		"""
		Short数组转Float数组（归一化到[-1, 1]）

		Args:
			data: Short数组

		Returns:
			Float数组
		"""
		return data.astype(np.float32) / 32768.0

	@staticmethod
	def teager_envelope(signal: np.ndarray) -> np.ndarray:
		"""
		Teager-Kaiser包络提取
		使用公式: Ψ[x(n)] = x²(n) - x(n-1)·x(n+1)
		返回包络信号 = sqrt(|Ψ|)

		Args:
			signal: 输入信号

		Returns:
			包络信号
		"""
		envelope = np.zeros(len(signal), dtype=np.float32)
		for i in range(1, len(signal) - 1):
			psi = signal[i] * signal[i] - signal[i - 1] * signal[i + 1]
			envelope[i] = math.sqrt(abs(psi))
		return envelope

	@staticmethod
	def calculate_teager_feature(envelope: np.ndarray) -> float:
		"""
		计算Teager包络特征值（平均幅度）

		Args:
			envelope: 包络信号

		Returns:
			平均幅度
		"""
		if len(envelope) == 0:
			return 0.0
		if len(envelope) <= 2:
			return 0.0
		return float(np.mean(envelope[1:-1]))
